### DISCUSSION TITLE
Pagination

### DISCUSSION THREAD

**Discussion by jakenoble** [CBIIT/ccdi-federation-api#15] [API Design Discussion]
## How Should We Approach Pagination?

St. Jude currently returns 100 rows per page and has approximately 120,000 rows in total.

### Options for Improvement

1. **Custom Rows Per Page**: We could enhance the API to accept a `rows per page` parameter, allowing end-users to customize the number of rows returned per page.
  
2. **Fetch All Rows**: Is there a specific need to enable fetching all rows in a single API request?

**Comment by asafieva** [comment]
1. This is a suggestion of a paginated result component specification.

```yaml
components:
  schemas:
    # components here
    responses.PaginatedResult:
      type: object
      properties:
        hasNext:
          type: boolean
          example: true
        hasPrev:
          type: boolean
          example: true
        page:
          type: number
          example: 3
        pages:
          type: number
          example: 5
        perPage:
          type: number
          example: 20
        total:
          type: number
          example: 100
        results:
          type: array
          items:
            oneOf:
              - $ref: '#/components/schemas/responses.Subject'
              # all components allowing pagination references here
```
2. Fetch All Rows can be useful for a full data load clients.

**Comment by jakenoble** [comment]
Apologies @asafieva, St Jude at present does have a `meta` portion of the response as

```
{
   "current_page":1,
   "per_page":100,
   "total":5594,
   "last_page":56,
   "from":1,
   "to":100
}
```

**Comment by e-t-k** [comment]
I don't see a real need for a custom-rows-per-page parameter.
Fetch-all-rows could be useful to enable users to avoid dealing with pagination (especially when a filtered results in relatively fewer rows being returned); but eg with St Jude's 120k rows it might cause performance issues on the client end if they're not expecting such a large response.
